# Panagram: Interactive, alignment-free pan-genome browser  

#### Katie Jenike, Sam Kovaka, Shujun Ou, Stephen Hwang, Srividya Ramakrishnan, Ben Langmead, Zach Lippman, Michael Schatz


An alignment-free pan-genome viewer

# Installation

```
> pip install git+https://github.com/kjenike/panagram.git --user
```
OR
```
> git clone --recursive git+https://github.com/kjenike/panagram.git
> cd panagram
> pip install .
```

# Preprocessing
Usage:
```
./panagram/scripts/preprocess.sh <genome_fastas.tsv> <K> <index_dir>
```

Where `<genome_fastas.tsv>` is a tab-separated file with the first column being a genome ID containing only alphanumeric characters, and the second a path to a gzipped fasta file. Will output all files to directory named `<index_dir>/`.

**TODO:**
```
preprocessing for genes, repeats, and MASH
config file format for "panagram view"
```

# View

Usage:
```
usage: panagram view [-h] <config>.txt
  Display panagram viewer
```

Runs a local Dash server. Browser can be viewed at http://127.0.0.1:8050/ by default.

# Bitdump

Usage:
```
usage: panagram bitdump [-h] [-v bool] str str str
  Query pan-kmer bitmap generated by "panagram index" 

  index_dir             Panagram index directory
  coords                Coordinates to query in chr:start-end format
  step                  Spacing between output kmers (optimized for multiples
                        of 100) (default: 1)
  -v bool, --verbose bool
                        Output the full bitmap (default: False)
```


## More information coming soon!